core/app/models/spree/tax/application.rb

```diff
@@ -0,0 +1,51 @@
+module Spree
+  module Tax
+    class Application
+      def initialize(order, output)
+        @order = order
+        @output = output
+      end
+
+      def perform!
+        %i(line_items shipments).each do |items|
+          order.send(items).each do |item|
+            rates = output.send(items).select { |i| i.id == item.id }
```

Comment:
I think we could avoid the `send`s here (but the rest of the structure looks good).